File: Binary Tree Postorder Traversal.py
# https://leetcode.com/problems/binary-tree-postorder-traversal/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    
    def postorder(self, root, nodes):
        
        # If root is empty, exit function
        if root is None:
            return
        
        # Visit children
        self.postorder(root.left, nodes)
        self.postorder(root.right, nodes)
        
        # Add root to list
        nodes.append(root.val)
    
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        
        # OBJECTIVE: Traverse tree in postorder order (left, right, root)
        
        # If root is empty, return an empty list
        if root is None:
            return list()
        
        # Create a list
        nodes = list()
        
        # Traverse tree
        self.postorder(root, nodes)
        return nodes

File: Binary Tree Postorder Traversal.java
// https://leetcode.com/problems/binary-tree-postorder-traversal/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    
    public void postorder(TreeNode root, List<Integer> nodes) {
        
        // If root is null, exit function
        if (root == null) {return;}
        
        // Visit children first
        postorder(root.left, nodes);
        postorder(root.right, nodes);
        
        // Add root node to list
        nodes.add(root.val);
    }
    
    public List<Integer> postorderTraversal(TreeNode root) {
        
        // OBJECTIVE: Traverse tree in postorder (left, right, root)
        
        // Create a list
        List<Integer> nodes = new ArrayList<Integer>();
        
        // If root is null, return an empty list
        if (root == null) {return nodes;}
        
        // Traverse tree
        postorder(root, nodes);
        return nodes;
    }
}
